conu/apidefs/filesystem.py

```diff
@@ -0,0 +1,141 @@
+import logging
+import os
+import shutil
+
+from tempfile import mkdtemp
+
+import xattr
+
+
+logger = logging.getLogger(__name__)
+
+
+class Filesystem(object):
+    """
+    Utility methods used to access filesystem of containers and images.
+
+    Implementations should probably be done using context managers.
+    """
+    def __init__(self, object_instance, mount_point=None):
+        """
+        :param object_instance: instance of the container or image
+        :param mount_point: str, directory where the filesystem will be mounted
+        """
+        self.obj = object_instance
+        self._mount_point = mount_point
+        self.mount_point_provided = False
+
+    @property
+    def mount_point(self):
+        if self._mount_point is None:
+            self._mount_point = mkdtemp(prefix="conu")
+            self.mount_point_provided = True
+        return self._mount_point
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.mount_point_provided:
+            os.rmdir(self.mount_point)
+
+    def p(self, path):
+        """
+        provide absolute path within the container
+
+        :param path: path with container
+        :return: str
+        """
+        if path.startswith("/"):
+            path = path[1:]
+        p = os.path.join(self.mount_point, path)
+        logger.debug("path = %s", p)
+        return p
+
+    def copy_to(self, src, dest):
+        """
+        copy a file or a directory from host system to a container -- don't implement for images,
+        those are immutable
+
+        :param src: str, path to a file or a directory on host system
+        :param dest: str, path to a file or a directory within container
+        :return: None
+        """
+        raise NotImplementedError("copy_to method is not implemented")
+
+    def copy_from(self, src, dest):
+        """
+        copy a file or a directory from container or image to host system
+
+        :param src: str, path to a file or a directory within container or image
+        :param dest: str, path to a file or a directory on host system
+        :return: None
+        """
+        p = self.p(src)
+        if os.path.isfile(src):
+            logger.info("copying file %s to %s", p, dest)
+            shutil.copy2(p, dest)
+        else:
+            logger.info("copying directory %s to %s", p, dest)
+            shutil.copytree(p, dest)
+
+    def read_file(self, file_path):
+        """
+        read file specified via 'file_path' and return its content - raises an ConuException if
+        there is an issue with read the file
+
+        :param file_path: str, path to the file to read
+        :return: str (not bytes), content of the file
+        """
+        with open(self.p(file_path)) as fd:
+            return fd.read()
+
+    def get_file(self, file_path, mode="r"):
```

Comment:
What is the reason to have default mode="r"? I would prefer to have it directly in `open` function.
Do you plan to read file as 'rb' and why?

Reply:
I don't understand, please elaborate.